### PR TITLE
feat: [PC-13433]: add get Adjustment Events

### PR DESCRIPTION
**Pull request by piotrkazulak**
## Motivation

This change is a part of new Error Budget Adjustments management feature. We need to have a way to get/update/delete Events related to an Adjustments.

## Summary

Add new command to `sloctl` to 

- get Adjustment Events for adjustment and optional SLO filter
- delete Adjustment Events for adjustment
- update Adjustment Events for adjustment

## Testing

See added examples of usage

## Release Notes

New Past Adjustment functionality. Allow users to manage past events.

### CODE REVIEW

File: internal/budgetadjustments/flags.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import (
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"github.com/nobl9/sloctl/internal/csv"
+)
+
+const (
+	flagAdjustment = "adjustment-name"
```
Comment (kubaceg) [inline internal/budgetadjustments/flags.go:12]: maybe just name? we have `sloctl budgetadjustments --adjustment-name` for me adjustment in flag is uncessary duplication

File: internal/budgetadjustments/flags.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	flagFrom       = "from"
+	flagTo         = "to"
+	flagProject    = "project"
```
Comment (kubaceg) [inline internal/budgetadjustments/flags.go:15]: `slo-project`? project is related to SLO not to adjustment so it could be confusing.

File: internal/budgetadjustments/flags.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	cmd.Flags().StringVarP(storeIn, flagProject, "", "",
+		"Name of the project. Required when sloName is defined.")
+}
+
+func registerSloNameFlag(cmd *cobra.Command, storeIn *string) {
+	cmd.Flags().StringVarP(storeIn, flagSloName, "", "",
+		"Name of the SLO. Required when sloName is defined.")
+}
+
+type TimeValue struct{ time.Time }
```
[the remark this answers is not in the document]
Comment (kubaceg) [reply internal/budgetadjustments/flags.go:54]: Also could You consider move whole declaration of this type to the beginning of this file? Personally I like when types, consts and global variables are declared on the beginning. WDYT?

File: internal/budgetadjustments/flags.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+const (
+	timeLayout       = time.RFC3339
+	timeLayoutString = "RFC3339"
```
Comment (kubaceg) [inline internal/budgetadjustments/flags.go:58]: unused value

File: internal/budgetadjustments/get.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+func NewGetCmd(client *sdk.Client) *cobra.Command {
+	get := &GetCmd{out: os.Stdout}
+
+	cmd := &cobra.Command{
+		Use:   "get",
+		Short: "Return a list of events for given Adjustment with related SLOs.",
+		Long: "This endpoint only return past and ongoing events (events that are already started)." +
+			"Please see Editing budget adjustments." +
+			"Maximum 500 events can be returned." +
```
Comment (kubaceg) [inline internal/budgetadjustments/get.go:53]: 500? not 250?

File: internal/budgetadjustments/get.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+			"Optional filtering for specific SLO (only one). If SLO is defined we will return only events" +
+			" for that SLO and the result will also include other SLOs that this events have. Sorted by eventStart.",
```
Comment (kubaceg) [inline internal/budgetadjustments/get.go:55]: Maybe `Sorted from oldest events to newest one.`

File: internal/budgetadjustments/get.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+func NewGetCmd(client *sdk.Client) *cobra.Command {
+	get := &GetCmd{out: os.Stdout}
+
+	cmd := &cobra.Command{
+		Use:   "get",
+		Short: "Return a list of events for given Adjustment with related SLOs.",
+		Long: "This endpoint only return past and ongoing events (events that are already started)." +
+			"Please see Editing budget adjustments." +
+			"Maximum 500 events can be returned." +
+			"Optional filtering for specific SLO (only one). If SLO is defined we will return only events" +
```
Comment (kubaceg) [inline internal/budgetadjustments/get.go:54]: Optional filtering for specific SLO (only one), in such cases both slo-name and slo-project flags are required.

File: internal/root.go [hunk 8]
```diff
@@ -54,6 +55,7 @@ For every command more detailed help is available.`,
 	rootCmd.AddCommand(root.NewConfigCmd())
 	rootCmd.AddCommand(root.NewReplayCmd())
 	rootCmd.AddCommand(root.NewAwsIamIds())
+	rootCmd.AddCommand(budgetadjustments.NewRootCmd(root.GetClient()))
```
Comment (kubaceg) [inline internal/root.go:58]: enter :p

**Review by kubaceg — CHANGES_REQUESTED** [review]

File: internal/budgetadjustments/root.go [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+	"fmt"
+
+	"github.com/nobl9/nobl9-go/sdk"
+	"github.com/spf13/cobra"
+)
+
+func NewRootCmd(client *sdk.Client) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "budgetadjustments",
+		Short: "Manage budgetadjustment.",
```
Comment (kubaceg) [inline internal/budgetadjustments/root.go:14]: `Budget adjustments managment` without dot at the end

File: internal/budgetadjustments/flags.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+func (t *TimeValue) String() string {
+	if t.IsZero() {
+		return ""
+	}
+	return t.Format(timeLayout)
+}
+
+func (t *TimeValue) Set(s string) (err error) {
+	t.Time, err = time.Parse(timeLayout, s)
```
Comment (kubaceg) [inline internal/budgetadjustments/flags.go:69]: You should catch error and throw own like 'Date in invalid format' because now it's golang's error like ''parsing time "budget-adjustment-recurr-2" as "2006-01-02T15:04:05Z07:00": cannot parse "budget-adjustment-recurr-2" as "2006" I know what's going on but this golang date format date can be unreadable for users.

File: internal/budgetadjustments/get.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+	if err := g.printObjects(events); err != nil {
+		return errors.Wrap(err, "failed to print objects")
+	}
+
+	return nil
+}
+
+func (g *GetCmd) printObjects(objects interface{}) error {
+	p, err := printer.New(
```
Comment (kubaceg) [inline internal/budgetadjustments/get.go:118]: what about printing errors? now I see json error from API 
```
Error: failed to get: bad response (status: 400): {"errors":[{"propertyName":"to","propertyValue":"{\"Adjustment\":\"budget-adjustment-recurr-2\",\"From\":\"2024-08-12T07:35:00Z\",\"To\":\"2024-08-12T07:35:00Z\"...","title":"'to' date must be be after 'from' date."}]}
```

**Review by kubaceg — CHANGES_REQUESTED** [review]

File: internal/budgetadjustments/flags.go (hunk 1 above)
Comment (piotrkazulak) [reply internal/budgetadjustments/flags.go]: We are getting events by adjustment name so lets leave it as it is

File: internal/budgetadjustments/flags.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (piotrkazulak) [reply internal/budgetadjustments/flags.go]: moved to separate file

File: internal/budgetadjustments/flags.go (hunk 4 above)
Comment (piotrkazulak) [reply internal/budgetadjustments/flags.go]: removed

File: internal/budgetadjustments/get.go (hunk 5 above)
Comment (piotrkazulak) [reply internal/budgetadjustments/get.go]: removed

File: internal/budgetadjustments/get.go (hunk 6 above)
Comment (piotrkazulak) [reply internal/budgetadjustments/get.go]: lets update it when doc will be ready

File: internal/budgetadjustments/get.go (hunk 7 above)
Comment (piotrkazulak) [reply internal/budgetadjustments/get.go]: lets update it when doc will be ready

File: internal/root.go (hunk 8 above)
Comment (piotrkazulak) [reply internal/root.go]: done

File: internal/budgetadjustments/root.go (hunk 9 above)
Comment (piotrkazulak) [reply internal/budgetadjustments/root.go]: done

File: internal/budgetadjustments/flags.go (hunk 10 above)
Comment (piotrkazulak) [reply internal/budgetadjustments/flags.go]: done, changed to:
```
Error: invalid argument "2023-09-23T20:46:00Za" for "--to" flag: to does not match 2006-01-02T15:04:05Z07:00 format
```

File: internal/budgetadjustments/get.go (hunk 11 above)
Comment (piotrkazulak) [reply internal/budgetadjustments/get.go]: switched to sdk.HTTPErors with sdk.APIErrors

File: internal/budgetadjustments/events/events.go [hunk 12]
```diff
@@ -0,0 +1,21 @@
+package events
+
+import (
+	_ "embed"
```
Comment (nieomylnieja) [inline internal/budgetadjustments/events/events.go:4]: redundant embed

File: internal/budgetadjustments/events/get.go [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+
+	resBody, err := request.DoRequest(
+		g.client,
+		cmd.Context(),
+		http.MethodGet,
+		fmt.Sprintf("%s/%s/events", request.BudgetAdjustmentAPI, g.adjustment),
+		values,
+	)
+	if err != nil {
+		return errors.Wrap(err, "failed to get events")
```
Comment (nieomylnieja) [inline internal/budgetadjustments/events/get.go:113]: you should not wrap this error, you're using `sdk.client.Do` under the hood which returns a well formatted error, ready to be printed, wrapping it will make it look worse, see https://github.com/nobl9/nobl9-go/blob/24782da0f04759accc905bfc73f4886c79623939/sdk/client.go#L140 for more details

File: internal/budgetadjustments/flags/flags.go [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+	outputFormat, fieldSeparator, recordSeparator *string,
+) {
+	cmd.PersistentFlags().StringVarP(outputFormat, "output", "o", "yaml",
+		`Output format: one of yaml|json|csv.`)
+
+	cmd.PersistentFlags().StringVarP(fieldSeparator, csv.FieldSeparatorFlag, "",
+		csv.DefaultFieldSeparator, "Field Separator for CSV.")
+
+	cmd.PersistentFlags().StringVarP(recordSeparator, csv.RecordSeparatorFlag, "",
+		csv.DefaultRecordSeparator, "Record Separator for CSV.")
```
Comment (nieomylnieja) [inline internal/budgetadjustments/flags/flags.go:52]: we shouldn't be redeclaring flags which are already used in sloctl, see https://github.com/nobl9/sloctl/blob/main/internal/flags.go. Not all of these flags that are already defined there should be shared, but some do, maybe we can extract these to a separate package?

**Review by nieomylnieja — CHANGES_REQUESTED** [review]

File: internal/budgetadjustments/events/events.go (hunk 12 above)
Comment (piotrkazulak) [reply internal/budgetadjustments/events/events.go:4]: Yes, I will remove it. BTW this is cool name for a band "Redundant embed" :guitar:

File: internal/budgetadjustments/root.go [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+
+	"github.com/spf13/cobra"
+
+	"github.com/nobl9/sloctl/internal/budgetadjustments/events"
+	"github.com/nobl9/sloctl/internal/sdkclient"
+)
+
+func NewRootCmd(clientProvider sdkclient.SdkClientProvider) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "budgetadjustments",
```
[the remark this answers is not in the document]
Comment (piotrkazulak) [reply internal/budgetadjustments/root.go:15]: short answer is no :(

File: internal/budgetadjustments/root.go [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+		Short: "Budget adjustments management",
+	}
+	cmd.PersistentFlags().BoolP("help", "h", false, fmt.Sprintf("Help for %s.", cmd.Name()))
+	cmd.AddCommand(events.NewRootCmd(clientProvider))
+	return cmd
```
[the remark this answers is not in the document]
Comment (piotrkazulak) [reply internal/budgetadjustments/root.go:20]: fair enough. I just had in my mind great plans for extending this feature. 

Yes, now it is not necessary. I will clean it up.

File: internal/budgetadjustments/events/inputreader.go [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+import (
+	"io"
+	"os"
+)
+
+func read(path string) (data []byte, err error) {
+	if path == "" || path == "-" {
+		return io.ReadAll(os.Stdin)
+	}
+	return os.ReadFile(path) // #nosec G304
```
Comment (nieomylnieja) [inline internal/budgetadjustments/events/inputreader.go:12]:
```suggestion
	if path == "" || path == "-" {
		return io.ReadAll(os.Stdin)
	}
        path = filepath.Clean(path)
	return os.ReadFile(path) // #nosec G304
```

File: internal/budgetadjustments/events/delete.go [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+
+//go:embed examples/delete_example.sh
+var deleteExample string
+
+func NewDeleteCmd(clientProvider sdkclient.SdkClientProvider) *cobra.Command {
+	deleteCmd := &DeleteCmd{out: os.Stdout}
+
+	cmd := &cobra.Command{
+		Use:     "delete",
+		Short:   "Delete existing events with new values. Values for eventStart and eventEnd are required.",
```
Comment (nieomylnieja) [inline internal/budgetadjustments/events/delete.go:40]: I'm not sure I understand this description, can we frame it differently? What does it mean that I "delete existing events with new values"? What new values :P

File: internal/budgetadjustments/events/flags.go [hunk 19]
```diff
@@ -17,7 +17,7 @@ const (
 
 func MustRegisterFileFlag(cmd *cobra.Command, storeIn *string) {
```
Comment (nieomylnieja) [inline internal/budgetadjustments/events/flags.go:18]: no need to export these flags from this file

File: go.mod [hunk 20]
```diff
@@ -12,6 +12,7 @@ require (
 	github.com/spf13/cobra v1.8.1
 	github.com/stretchr/testify v1.9.0
 	golang.org/x/sync v0.8.0
+	gopkg.in/yaml.v3 v3.0.1
```
Comment (nieomylnieja) [inline go.mod:15]: you've used the wrong yaml library ;) we're using https://github.com/goccy/go-yaml

File: internal/budgetadjustments/events/delete.go [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+	var yamlData []Event
+	if err := yaml.Unmarshal(data, &yamlData); err != nil {
+		return errors.Wrap(err, "failed to load input data")
+	}
+	jsonData, err := json.Marshal(yamlData)
+	if err != nil {
+		return errors.Wrap(err, "failed to convert input data to JSON")
+	}
+
+	if g.dryRun {
```
Comment (nieomylnieja) [inline internal/budgetadjustments/events/delete.go:72]: I'm not sure we should be using `--dry-run` flag here, it will be confusing in the context of other commands as it doesn't run the request in `dry-run` mode, instead maybe use `validate`? (although it seems we're not validating anything here)

File: test/adjustments.bats [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+  load "test_helper/load"
+  load_lib "bats-support"
+  load_lib "bats-assert"
+}
+
+@test "missing required flags" {
+	run_sloctl budgetadjustments events get
+  assert_failure
+  output="$stderr"
+  assert_output 'Error: required flag(s) "adjustment-name", "from", "to" not set'
```
Comment (nieomylnieja) [inline test/adjustments.bats:23]: some formatting issues here:
image

**Review by nieomylnieja — CHANGES_REQUESTED** [review]

File: internal/budgetadjustments/events/examples/delete_example.sh [hunk 23]
```diff
@@ -1,11 +1,26 @@
-# Delete Adjustment Evens using a file
-sloctl budgetadjustments events delete --adjustment-name=sample-adjustment-name -f /path/to/file/with/events
+# Delete Adjustment Events using a file
+sloctl budgetadjustments events delete --adjustment-name=sample-adjustment-name -f /path/to/file/with/events/yaml
```
Comment (nieomylnieja) [inline internal/budgetadjustments/events/examples/delete_example.sh:2]: you can delete it since you already have that at line 13 :)

File: internal/budgetadjustments/events/examples/update_example.sh [hunk 24]
```diff
@@ -1,7 +1,7 @@
-# Update Adjustment Evens using a file
+# Update Adjustment Events using a file
```
Comment (nieomylnieja) [inline internal/budgetadjustments/events/examples/update_example.sh:1]: you can delete it since you already have that at line 18 :)

File: internal/budgetadjustments/events/update.go [hunk 25]
```diff
@@ -62,9 +56,6 @@ func (g *UpdateCmd) run(cmd *cobra.Command) error {
 	if err != nil {
```
Comment (nieomylnieja) [inline internal/budgetadjustments/events/update.go:56]: you can use `yaml.YAMLToJSON()` function here ;) same goes for delete

File: internal/replay.go [hunk 26]
```diff
@@ -104,15 +104,8 @@ func (r *ReplayCmd) RunReplays(cmd *cobra.Command, replays []ReplayConfig) (fail
...
-			len(replays),
-			replay.SLO,
-			replay.Project,
-			replay.From.Format(
-				flags.TimeLayout,
-			),
-			time.Now().In(replay.From.Location()).Format(flags.TimeLayout),
-		)))
+			i+1, len(replays), replay.SLO, replay.Project,
+			replay.From.Format(flags.TimeLayout), time.Now().In(replay.From.Location()).Format(flags.TimeLayout))))
```
Comment (nieomylnieja) [inline internal/replay.go:108]: why did you change the formatting? please revert this

File: internal/replay.go [hunk 27]
```diff
@@ -528,16 +521,8 @@ func (r *ReplayCmd) replayUnavailabilityReasonExplanation(
...
-			expectedDuration.String(),
-			timeNow.Format(
-				flags.TimeLayout,
-			),
-			r.from.Format(flags.TimeLayout),
-			startOffsetMinutes,
-			offsetNotice,
-		)
+			replay.metricSource.Name, replay.metricSource.Project, expectedDuration.String(),
+			timeNow.Format(flags.TimeLayout), r.from.Format(flags.TimeLayout), startOffsetMinutes, offsetNotice)
```
Comment (nieomylnieja) [inline internal/replay.go:525]: why did you change the formatting? please revert this

File: internal/budgetadjustments/events/examples/delete_example.sh [hunk 28]
```diff
@@ -1,11 +1,26 @@
...
 
-# Delete Adjustment Event using stdin:
+# Delete Adjustment Events using stdin:
 echo '
```
Comment (nieomylnieja) [inline internal/budgetadjustments/events/examples/delete_example.sh:5]: to make this whole example file more concise if you move the below example up:
same goes for update

```sh
# Delete Adjustment Events using a file:
cat <<EOF > ./events.yaml
- eventStart: 2024-10-24T04:07:04Z
  eventEnd: 2024-10-24T05:27:04Z
  slos:
  - project: test-project
    name: sample-slo-1
- eventStart: 2024-10-25T04:07:04Z
  eventEnd: 2024-10-25T05:27:04Z
  slos:
  - project: test-project
    name: sample-slo-2
EOF
sloctl budgetadjustments events delete --adjustment-name=sample-adjustment-name -f ./events.yaml

# Delete Adjustment Events using stdin:
sloctl budgetadjustments events delete --adjustment-name=sample-adjustment-name -f - <./events.yaml
```

**Review by nieomylnieja — CHANGES_REQUESTED** [review]

File: internal/budgetadjustments/events/examples/delete_example.sh (hunk 28 above)
Comment (piotrkazulak) [reply internal/budgetadjustments/events/examples/delete_example.sh]: rearranged

File: internal/replay.go (hunk 27 above)
Comment (piotrkazulak) [reply internal/replay.go:525]: No, I didn't :confused:

File: internal/replay.go (hunk 26 above)
Comment (piotrkazulak) [reply internal/replay.go:108]: No, I didn't :confused:

File: internal/budgetadjustments/events/update.go (hunk 25 above)
Comment (piotrkazulak) [reply internal/budgetadjustments/events/update.go:51]: OK

File: internal/budgetadjustments/events/examples/update_example.sh (hunk 24 above)
Comment (piotrkazulak) [reply internal/budgetadjustments/events/examples/update_example.sh]: rearranged

File: internal/budgetadjustments/events/examples/delete_example.sh (hunk 23 above)
Comment (piotrkazulak) [reply internal/budgetadjustments/events/examples/delete_example.sh]: rearranged

File: internal/budgetadjustments/events/update.go [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+
+//go:embed examples/update_example.sh
+var updateExample string
+
+func NewUpdateCmd(clientProvider sdkclient.SdkClientProvider) *cobra.Command {
+	update := &UpdateCmd{}
+
+	cmd := &cobra.Command{
+		Use:     "update",
+		Short:   "Update existing events with new values. Values for eventStart and eventEnd are required.",
```
Comment (klaudia-nobl9) [inline internal/budgetadjustments/events/update.go:31]:
```suggestion
		Short:   "Update past existing events with new values. Values for eventStart and eventEnd are required.",
```

File: internal/budgetadjustments/events/delete.go [hunk 30]
```diff
@@ -0,0 +1,67 @@
...
+
+//go:embed examples/delete_example.sh
+var deleteExample string
+
+func NewDeleteCmd(clientProvider sdkclient.SdkClientProvider) *cobra.Command {
+	deleteCmd := &DeleteCmd{}
+
+	cmd := &cobra.Command{
+		Use:     "delete",
+		Short:   "Delete existing events.",
```
Comment (klaudia-nobl9) [inline internal/budgetadjustments/events/delete.go:31]:
```suggestion
		Short:   "Delete past existing events.",
```

**Review by kubaceg — APPROVED** [review]
